lib/extensions.js

```diff
@@ -106,38 +113,112 @@ module.exports = function(realmConstructor, context) {
+            let behavior = realmConstructor.exists(config) ? "existingRealmFileBehavior" : "newRealmFileBehavior";
+            
+            // Define how the Realm file is opened
+            let openLocalRealmImmediately = false; // Default is syncBeforeOpen
+            if (config.sync[behavior] !== undefined) {
+                const type = config.sync[behavior].type;
+                switch (type) {
+                    case 'syncBeforeOpen': 
+                        openLocalRealmImmediately = false; 
+                        break;
+                    case 'openImmediately': 
+                        openLocalRealmImmediately = true;
+                        break;
+                    default:
+                        throw Error("Invalid type: '" + type + "'. Only 'syncBeforeOpen' and 'openImmediately' is allowed.")
+                }
+            }
+
+            // Configure timeout
+            let timeOut = 0x7FFFFFFF; // Max value allowed by setTimeOut()
+            if (config.sync[behavior] !== undefined && config.sync[behavior].timeOut !== undefined) {
+                let userDefinedTimeOut = config.sync[behavior].timeOut;
+                if (typeof userDefinedTimeOut !== 'number') {
+                    throw new Error("'timeOut' must be a number: '" + userDefinedTimeOut + "'");
+                }
+                timeOut = userDefinedTimeOut;
+            }
 
-            promise.progress = (callback) => {
-                if (syncSession) {
-                    syncSession.addProgressNotification('download', 'forCurrentlyOutstandingWork', callback);
+            // Define the behavior in case of a timeout
+            let throwOnTimeOut = true; // Default is to throw
+            if (config.sync[behavior] !== undefined && config.sync[behavior].timeOutBehavior) {
+                const timeOutBehavior = config.sync[behavior].timeOutBehavior;
+                switch (timeOutBehavior) {
+                    case 'throwException': 
+                        throwOnTimeOut = true;
+                        break;
+                    case 'openLocal':
+                        throwOnTimeOut = false;
+                        break;
+                    default:
+                        throw Error("Invalid 'timeOutBehavior': '" + timeOutBehavior + "'. Only 'throwException' and 'openLocal' is allowed.")
                 }
+            }
 
-                return promise;
-            };
-            return promise;
+            // Opens the synchronized Realm
+            if (openLocalRealmImmediately) {
+                return openLocalRealm(realmConstructor, config);
+            } else {
+                let asyncOpenTask;
+                let openRemoteRealmPromise = new Promise((resolve, reject) => {
+                    asyncOpenTask = realmConstructor._asyncOpen(config, (realm, error) => {
+                        if (error) {
+                            setTimeout(() => {
```

Comment:
Pushing the `if (error)` check inside the timeout function cuts down on duplicated code.